### ISSUE TITLE
Add SPSA optimization method

### ISSUE THREAD

**Issue by ankit27kh** [PennyLaneAI/pennylane#2451]
### Feature details

The  Simultaneous Perturbation Stochastic Approximation (SPSA) optimisation method is a faster optimisation method.

>  If the number of terms being optimized is p, then the finite-difference method takes 2p measurements of the objective function at each iteration (to form one gradient approximation), while SPSA takes only two measurements

It is also naturally suited for noisy measurements. Thus, it will be useful when simulating noisy systems.

The theory (and implementation) for SPSA is:
* Described in James C. Spall, “[An Overview of the Simultaneous Perturbation Method for Efficient Optimization](https://www.jhuapl.edu/SPSA/PDF-SPSA/Spall_An_Overview.PDF)”, 1998;
* Showcased in the [Optimization using SPSA](https://pennylane.ai/qml/demos/spsa.html) PennyLane tutorial.

Furthermore, it is implemented:
* In the [`noisyopt` package](https://noisyopt.readthedocs.io/en/latest/), specifically see [the source code here](https://github.com/andim/noisyopt/blob/master/noisyopt/main.py#L264);
* In Qiskit, [see the SPSA optimizer documentation](https://qiskit.org/documentation/stubs/qiskit.algorithms.optimizers.SPSA.html#qiskit.algorithms.optimizers.SPSA).

#### 1. Implementation

There exists a `qml.SPSAOptimizer` optimizer that:
* Implements the `compute_grad` and the `apply_grad` methods ([see `GradientDescentOptimizer.apply_grad` doc](https://pennylane.readthedocs.io/en/stable/code/api/pennylane.GradientDescentOptimizer.html?highlight=gradientdescent#pennylane.GradientDescentOptimizer.apply_grad) and [`GradientDescentOptimizer.compute_grad` doc](https://pennylane.readthedocs.io/en/stable/code/api/pennylane.GradientDescentOptimizer.html?highlight=gradientdescent#pennylane.GradientDescentOptimizer.compute_grad));
* Implements a `step` method ([see `GradientDescentOptimizer.step` doc](https://pennylane.readthedocs.io/en/stable/code/api/pennylane.GradientDescentOptimizer.html?highlight=gradientdescent#pennylane.GradientDescentOptimizer.step));
* Implements a `step_and_cost` method ([see `GradientDescentOptimizer.step_and_cost` doc](https://pennylane.readthedocs.io/en/stable/code/api/pennylane.GradientDescentOptimizer.html?highlight=gradientdescent#pennylane.GradientDescentOptimizer.step_and_cost));
* The `SPSAOptimizer` allows passing values to the quantities that greatly influence the optimization, including:
    * `a`: scaling parameter for step size;
    * `A`: stability constant;
    * `c`: scaling parameter for evaluation step size;
    * `alpha`: scaling exponent for step size;
    * `gamma`: scaling exponent for evaluation step size;
    but also defines defaults suggested in the paper. Such defaults may be of the most practical use in many use cases. E.g., `noisyopt` also uses such defaults, except for `A` that is "fixed" based on the number of iteration steps.
#### 2. Testing

Test the `SPSAOptimizer` class similar to how [`GradientDescentOptimizer`](https://github.com/PennyLaneAI/pennylane/blob/master/tests/optimize/test_gradient_descent.py) and [`QNGOptimizer`](https://github.com/PennyLaneAI/pennylane/blob/master/tests/optimize/test_qng.py) are being tested.

#### 3. Documentation

The documentation of the optimizer should be similar in fashion to the [documentation of the `QNGOptimizer`](https://pennylane.readthedocs.io/en/stable/code/api/pennylane.QNGOptimizer.html?highlight=qngoptimizer#pennylane.QNGOptimizer).

The docstring of `qml.SPSAOptimizer` should:
* Introduce the concept of SPSA and the task it solves;
* Include and explain equations (5) and (6) from “[An Overview of the Simultaneous Perturbation Method for Efficient Optimization](https://www.jhuapl.edu/SPSA/PDF-SPSA/Spall_An_Overview.PDF)” (page 9 of the PDF that is page 490 of the volume);
* Mention the number of quantum device executions required with SPSA;
* Mention the use cases for SPSA as the optimization method of a hybrid quantum-classical workflow;
* Add examples for showcasing the use of the optimizer as for `QNGOptimizer`.

### How important would you say this feature is?

2: Somewhat important. Needed this quarter.

### Additional information

https://www.jhuapl.edu/SPSA/

**Comment by antalszava** [comment]
Hi @ankit27kh, thank you! We have a [demonstration showcasing how to use PennyLane with SPSA optimization](https://pennylane.ai/qml/demos/spsa.html). In the demo, the `noisyopt` package is used as it integrates well with PennyLane.

Would it be beneficial if PennyLane came with a dedicated SPSA optimizer?

**Comment by ankit27kh** [comment]
Hi @antalszava, thanks for the link. I wasn't aware of this.
I would still suggest having SPSA built into PennyLane as it is advantageous for optimisations in the quantum world and should be available without needing to install additional libraries.
After integration, it should work like other optimisers:
```
opt = qml.SPSA()
for _ in range(10):
  params = opt.step(cost, params)
```

**Comment by ankit27kh** [comment]
What I like about PennyLane optimisers is the ability to have granular control over the iterations. This is usually missing from other libraries like `qiskit` and `scipy`. Sometimes you can use callback functions, but it's not the same.
I hope we can have SPSA in PennyLane so that we still enjoy a high degree of control over the optimisation process.

**Comment by antalszava** [comment]
Hi @ankit27kh, thank you so much for this feedback! In fact, we've been considering adding a dedicated SPSA optimizer to PennyLane. Good to hear that there would be value in such an addition. :+1:

**Comment by ixime** [comment]
I am not sure what is required for the use cases of hybrid quantum-classical workflow, because as far as I know, in those cases, the optimizers used are the ones of the classical platforms, i.e. [tf.keras.optimizers.Adam](https://www.tensorflow.org/api_docs/python/tf/keras/optimizers/Adam)

**Comment by antalszava** [comment]
Hi @ixime, the part on quantum classical workflows just means to mention in the docstring of the optimizer the fact that SPSA can be particularly beneficial when optimizing workflows that are using noisy quantum devices.

**Comment by maxwell04-wq** [comment]
Hi,

I would like to contribute a tutorial for simulated annealing using PennyLane. Can be converted in a module too. Can I do it under UnitaryHack?
